airflow/jobs.py

```diff
@@ -484,84 +484,99 @@ def process_dag(self, dag, queue):
             return None
-        else:
-            # Taking a lock
-            db_dag.scheduler_lock = True
-            db_dag.last_scheduler_run = datetime.now()
-            session.commit()
 
-        active_runs = dag.get_active_runs()
+        # Release the db lock
+        # the assumption here is that process_dag will take less
+        # time than self.heartrate otherwise we might unlock too
+        # quickly and this should moved below, but that would increase
+        # the time the record is locked and is blocking for other calls.
+        db_dag.last_scheduler_run = datetime.now()
+        session.commit()
 
-        self.logger.info('Getting list of tasks to skip for active runs.')
-        skip_tis = set()
-        if active_runs:
-            qry = (
-                session.query(TI.task_id, TI.execution_date)
-                .filter(
-                    TI.dag_id == dag.dag_id,
-                    TI.execution_date.in_(active_runs),
-                    TI.state.in_((State.RUNNING, State.SUCCESS, State.FAILED)),
-                )
-            )
-            skip_tis = {(ti[0], ti[1]) for ti in qry.all()}
+        # update the state of the previously active dag runs
+        dag_runs = DagRun.find(dag_id=dag.dag_id, state=State.RUNNING)
+        active_dag_runs = []
+        for run in dag_runs:
+            # do not consider runs that are executed in the future
+            if run.execution_date > datetime.now():
+                continue
 
-        descartes = [obj for obj in product(dag.tasks, active_runs)]
-        could_not_run = set()
-        self.logger.info('Checking dependencies on {} tasks instances, minus {} '
-                     'skippable ones'.format(len(descartes), len(skip_tis)))
+            # todo: run.task is transient but needs to be set
+            run.dag = dag
+            # todo: preferably the integrity check happens at dag collection time
+            run.verify_integrity()
+            run.update_state()
+            if run.state == State.RUNNING:
+                active_dag_runs.append(run)
 
-        for task, dttm in descartes:
-            if task.adhoc or (task.task_id, dttm) in skip_tis:
-                continue
-            ti = TI(task, dttm)
+        could_not_run = 0
+        available_tasks = 0
 
-            ti.refresh_from_db()
-            if ti.state in (
-                    State.RUNNING, State.QUEUED, State.SUCCESS, State.FAILED):
-                continue
-            elif ti.is_runnable(flag_upstream_failed=True):
-                self.logger.debug('Queuing task: {}'.format(ti))
-                queue.put((ti.key, pickle_id))
-            elif ti.is_premature():
-                continue
-            else:
-                self.logger.debug('Adding task: {} to the COULD_NOT_RUN set'.format(ti))
-                could_not_run.add(ti)
+        for run in active_dag_runs:
+            tis = run.get_task_instances(session=session, state=(State.NONE,
```

Comment:
As discussed on gitter. The check for deadlock here does not make sense anymore. TaskInstances are sent to the executor in a "SCHEDULED" state or in "UP_FOR_RETRY". The count of available tasks can therefore equal the amount of tasks that are not runnable as tasks can actually be running at that moment. "update_state" does check for deadlocked dag_runs by using all relevant states.